File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/ClusterSyncHandler.java
package com.sequenceiq.cloudbreak.reactor;

import jakarta.inject.Inject;

import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.domain.stack.Stack;
import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
import com.sequenceiq.cloudbreak.eventbus.Event;
import com.sequenceiq.cloudbreak.eventbus.EventBus;
import com.sequenceiq.cloudbreak.reactor.api.event.resource.ClusterSyncRequest;
import com.sequenceiq.cloudbreak.reactor.api.event.resource.ClusterSyncResult;
import com.sequenceiq.cloudbreak.service.cluster.ClusterService;
import com.sequenceiq.cloudbreak.service.cluster.InstanceMetadataUpdater;
import com.sequenceiq.cloudbreak.service.cluster.flow.status.ClusterStatusUpdater;
import com.sequenceiq.cloudbreak.service.image.userdata.UserDataService;
import com.sequenceiq.cloudbreak.service.stack.StackService;
import com.sequenceiq.cloudbreak.service.upgrade.sync.template.ClusterManagerTemplateSyncService;
import com.sequenceiq.flow.event.EventSelectorUtil;
import com.sequenceiq.flow.reactor.api.handler.EventHandler;

@Component
public class ClusterSyncHandler implements EventHandler<ClusterSyncRequest> {
    @Inject
    private StackService stackService;

    @Inject
    private ClusterService clusterService;

    @Inject
    private ClusterStatusUpdater clusterStatusUpdater;

    @Inject
    private EventBus eventBus;

    @Inject
    private InstanceMetadataUpdater instanceMetadataUpdater;

    @Inject
    private UserDataService userDataService;

    @Inject
    private ClusterManagerTemplateSyncService clusterManagerTemplateSyncService;

    @Override
    public String selector() {
        return EventSelectorUtil.selector(ClusterSyncRequest.class);
    }

    @Override
    public void accept(Event<ClusterSyncRequest> event) {
        ClusterSyncRequest request = event.getData();
        ClusterSyncResult result;
        try {
            Stack stack = stackService.getByIdWithListsInTransaction(request.getResourceId());
            Cluster cluster = clusterService.retrieveClusterByStackIdWithoutAuth(request.getResourceId()).orElse(null);
            clusterStatusUpdater.updateClusterStatus(stack, cluster);
            if (cluster != null && (stack.isAvailable() || stack.isMaintenanceModeEnabled())) {
                instanceMetadataUpdater.updatePackageVersionsOnAllInstances(stack.getId());
                userDataService.makeSureUserDataIsMigrated(stack.getId());
                clusterManagerTemplateSyncService.sync(stack.getId());
            }
            result = new ClusterSyncResult(request);
        } catch (Exception e) {
            result = new ClusterSyncResult(e.getMessage(), e, request);
        }
        eventBus.notify(result.selector(), new Event<>(event.getHeaders(), result));
    }

}

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/ClusterSyncHandler.kt
package com.sequenceiq.cloudbreak.reactor

import javax.inject.Inject

import org.springframework.stereotype.Component

import com.sequenceiq.cloudbreak.domain.Cluster
import com.sequenceiq.cloudbreak.domain.Stack
import com.sequenceiq.cloudbreak.reactor.api.event.resource.ClusterSyncRequest
import com.sequenceiq.cloudbreak.reactor.api.event.resource.ClusterSyncResult
import com.sequenceiq.cloudbreak.service.cluster.ClusterService
import com.sequenceiq.cloudbreak.service.cluster.flow.status.AmbariClusterStatusUpdater
import com.sequenceiq.cloudbreak.service.stack.StackService

import reactor.bus.Event
import reactor.bus.EventBus

@Component
class ClusterSyncHandler : ClusterEventHandler<ClusterSyncRequest> {
    @Inject
    private val stackService: StackService? = null
    @Inject
    private val clusterService: ClusterService? = null
    @Inject
    private val ambariClusterStatusUpdater: AmbariClusterStatusUpdater? = null
    @Inject
    private val eventBus: EventBus? = null

    override fun type(): Class<ClusterSyncRequest> {
        return ClusterSyncRequest::class.java
    }

    override fun accept(event: Event<ClusterSyncRequest>) {
        val request = event.data
        val result: ClusterSyncResult
        try {
            val stack = stackService!!.getById(request.stackId)
            val cluster = clusterService!!.retrieveClusterByStackId(request.stackId)
            ambariClusterStatusUpdater!!.updateClusterStatus(stack, cluster)
            result = ClusterSyncResult(request)
        } catch (e: Exception) {
            result = ClusterSyncResult(e.message, e, request)
        }

        eventBus!!.notify(result.selector(), Event(event.headers, result))
    }
}
